core/src/Streamly/Internal/Unicode/Stream.hs

```diff
@@ -965,13 +1114,52 @@ encodeUtf8Lax = encodeUtf8
 -- Encoding to Utf16
 -------------------------------------------------------------------------------
 
+{-# INLINE utf16LowSurrogate #-}
+utf16LowSurrogate :: Word16
+utf16LowSurrogate = 0xDC00
+
+{-# INLINE utf16HighSurrogate #-}
+utf16HighSurrogate :: Word16
+utf16HighSurrogate = 0xD800
+
+{-# INLINE_NORMAL readCharUtf16With #-}
+readCharUtf16With :: Monad m => WList Word16 -> Unfold m Char Word16
+readCharUtf16With invalidReplavement = Unfold step inject
+
+    where
+
+    inject c =
+        return $ case ord c of
+            x | x <= 0xD7FF -> fromIntegral x `WCons` WNil
```

Comment:
Should use the same numbers everywhere e.g. you used 0xD800 earlier, here also we should use `x < 0xD800`.